### DISCUSSION TITLE
O365 --> MS365 - A potential big change - your views needed

### DISCUSSION THREAD

**Discussion by RogerSelwyn** [RogerSelwyn/O365-HomeAssistant#234] [Ideas]
## 👋 Your views needed!
I'm using Discussions as a place to connect with other members of the community and understand your point of view on a potential major change.

I've noted in the release notes to [v4.8.0](https://github.com/RogerSelwyn/O365-HomeAssistant/releases/tag/v4.8.0) that the integration is becoming unwieldy. It has grown under my stewardship from an integration that supported [calendars and email](https://github.com/PTST/O365-HomeAssistant), to one that supports a whole variety of different sensors. This makes it slow to start up and operate and along with being a massive chunk of code to maintain.

I've started down a road that to breaking this up into smaller components, probably along the lines of `Calendar`, `Email` and `Teams`. To that end I have created a new integration just for calendars. I've branded this as `MS365` inline with Microsoft's changed branding of Office 365 to Microsoft 365. This has the benefit of clearly separating the integrations.

I'm very conscious that there is a significant following of this integration, so this may be an unpopular change, but I feel it is necessary. Other may suggest doing it another way, and I'm certainly open to views. My knowledge of what can be done with Python and HA is a long way of perfect, I'm an enthusiastic amateur.

I've taken to decision to make in the installation of the integration as being UI only. The calendars.yaml file still exists and the calendars can be managed via that method, but there is the ability to modify most of the config via the UI.

Again, this may be an unpopular decision because many want to be yaml only. I have previously managed an integration that does both YAML and UI, but it is a pain, so I'm not doing it again. I think with recent advances in the HA UI around reconfigure and reload, this will remove the need to go delete tokens or reboot HA.

Anyway, I'm open to comments. For reference the new calendar is here [MS365-Calendar](https://github.com/RogerSelwyn/MS365-Calendar). I would warn that it needs HA 2024.6.x since it incorporates changes included with that release.

**Comment by B-Hartley** [comment]
I only really use to-do
Where will that sit?

I have no strong feelings either way.

If it's easier for you to break it up. That's good enough for me.

**Reply by topgun0dk** [reply]
Same i do

**Reply by RogerSelwyn** [reply]
I have already created the new ToDo integration.

**Reply by topgun0dk** [reply]
I saw that further down. And it is now installed :)
Thanks a lot for your work

**Comment by RogerSelwyn** [comment]
Probably in Teams, but I may make that a fourth Integration. It’s quite complex like calendar.

**Reply by lexiismadd** [reply]
I think separated is a good idea, easier to manage, easier to diagnose issues.
Teams is definitely one area I use

**Reply by RogerSelwyn** [reply]
The dedicated Teams integration is available, there is a link in this thread further down

**Comment by SamBouwer** [comment]
@RogerSelwyn thank you for the open communication! I would fully support the split into separate integrations like you did for Calendar. It might be a (bit of a) pain for users, but I can see it benefits the maintenance of the integrations.

**Comment by uSlackr** [comment]
@RogerSelwyn I've been using this since you took it over and appreciate the long-term effort and commitment you've made to this project.  I like the idea of breaking up the code to reduce the load on the system and the less code the fewer bugs affect any particular install.  Are you concerned you could wind up making the same changes 3-4 times?  
FWIW I use calendar and to-do.

**Reply by RogerSelwyn** [reply]
I am concerned about duplication in the initial, config_flow, permissions logic. I might break it out into generic/specific code blocks, so I can copy content between integrations. I'll figure that as I get into the second one. I may break ToDo into a fourth module since it is similar to calendar in concept.

**Comment by RogerSelwyn** [comment]
As a passing comment, I haven't taken advantage of HA application credentials, because in general that seems to assume it will request, hold and manage the token, and the python-o365 simply isn't structured to work like that.

**Comment by loryanstrant** [comment]
I do think it's logical to break it up, as realistically there are different use cases across the M365 stack, plus it allows more people to contribute to certain areas of passion (and hopefully for those who have created other "competing" components to pool their efforts into these ones).
While I know practically nothing of Python, I do work with M365 and Graph daily (it's literally my job to help orgs out with this stuff and build custom solutions on it), so if you like I'm happy to get more involved across all the various components (whether it's architecture, integration in M365, functionality, testing, documentation, etc.).

**Reply by RogerSelwyn** [reply]
Thank you very much for your offer, much appreciated. I'll do the split up, and try to document what it is doing as best I can in some form (given it goes via a 3rd party library). Then maybe you can have a look and see if there are better ways of doing what it does.

**Comment by smarthomefamilyverrips** [comment]
First of all thank you for the work you put in to this. To me it seems a good choice as looking at other comments here more people just use a part of the functionality of the O365 integration same as I do. I only need/use it for getting Calendar data of my Outlook Calendar from work, so I personally I prefer a less heavy integration giving me just what I need. So definitely a +1 from me on this.

I surely will try out the M365-Calendar integration.

**Comment by RogerSelwyn** [comment]
For those interested, I have started writing up some documentation on what calls are made as part of the main retrieval cycle - https://github.com/RogerSelwyn/O365-HomeAssistant/wiki

I need to think about repository structure, to see if I can store all the integrations in one repository rather than multiple. This will depend on what HACS supports.

**Reply by RogerSelwyn** [reply]
> I need to think about repository structure, to see if I can store all the integrations in one repository rather than multiple. This will depend on what HACS supports.

Well that question is answered:

```
- There must only be one integration per repository, i.e. there can only be one subdirectory to ROOT_OF_THE_REPO/custom_components/. If there are more than one, only the first one will be managed.
- All files required for the integration to run must be located inside the directory ROOT_OF_THE_REPO/custom_components/INTEGRATION_NAME/.
```

**Comment by RogerSelwyn** [comment]
I've just released a first version of the [MS365-ToDo](https://github.com/RogerSelwyn/MS365-ToDo) integration. Again needs 2024.6.x

**Reply by B-Hartley** [reply]
It says I need 2024.6.0 to install it, but that version isn't out yet.
I'm on 2024.6.0b4

**Reply by B-Hartley** [reply]
I've installed manually for now, I'll let you know how it goes.

**Reply by RogerSelwyn** [reply]
I would have thought it would allow that. You can always manually copy it. Otherwise wait until Monday

**Reply by B-Hartley** [reply]
Installed manually, configured through UI.  All seemed to go ok and my lists are showing.  No time to test more right now, but looks ok !  Thanks

**Comment by RogerSelwyn** [comment]
I've just released a first version of the [MS365-Mail](https://github.com/RogerSelwyn/MS365-Mail) integration. Again needs 2024.6.x

Teams is still to come, but won't get done until mid-July. I've also requested HACS inclusion for the three already done, but that will take some months to get included.

**Comment by RogerSelwyn** [comment]
I've just release a first version of [MS365-Teams](https://github.com/RogerSelwyn/MS365-Teams). I will submit a HACS inclusion request as soon as the brand icon has been accepted into HA Core. 

Expect HACS to take many months, it is very, very slow.

**Comment by RogerSelwyn** [comment]
Just to update. HACS seems to have stopped accepting new integrations at the moment. I suspect this is because a new version is being developed which is currently in Alpha (or was when I looked a week ago). Hopefully when it is released they will work through the backlog.

I will endeavour to maintain both the O365 integration, and the split up versions until they make it onto HACS proper. I will then do some "marketing" on the new ones to encourage the move so I can decommission the old. This of course may take some time.

**Comment by tastyllama** [comment]
Thanks for the updates!  Appreciate all your hard work.

**Comment by cromulus** [comment]
Amazing work. Thank you!!! IMHO, this should be HA core.

**Reply by RogerSelwyn** [reply]
Thanks. Really appreciated. 

The work to get into HA Core is substantial, I’m just trying to work through creating the testing for another integration to maybe enable to get into core. Don’t expect anything soon.

**Reply by ClearlyDazed** [reply]
Thank you for this amazing work. Splitting up into individual modules is an excellent way forward. Not everyone utilizes all the modules from Microsoft 365. 
With the recent security upgrade by Microsoft to allow only Oauth2 your project seems to be a godsent to access Outlook 365 to send email notifications from Home Assistant. It also seems that Microsoft Azure has changed and your instructions on setting up Azure do not correlate. I was wondering if you could elaborate on what the new procedure would be to setup Azure.
Thanks

**Reply by RogerSelwyn** [reply]
To be honest I have no idea. Not sure I've ever received an email saying I must be on MFA by 14th October for access to Entra Portal (new name for Azure). Seems I might need a P1 license which is unlikely to ever happen, I only use it for email.

It might die from what I can see.

**Reply by RogerSelwyn** [reply]
> It also seems that Microsoft Azure has changed and your instructions on setting up Azure do not correlate. I was wondering if you could elaborate on what the new procedure would be to setup Azure. 

Not withstanding the MFA issue I mentioned, the create of the Entra/Azure app is exactly as described in the documentation. Perhaps you could expand on how you think it has changed?

**Reply by ClearlyDazed** [reply]
Your link to the portal initially sent me to an older site that required me to choose a site to register an account first. Everything else fell into place once I found the app registration option.

**Reply by RogerSelwyn** [reply]
Ok, I’ll check

**Reply by leedrake** [reply]
I do have a P1 license so if you need me to test something let me know.  I also have a tenant with a windows premium account which includes a P2 license.

**Reply by leedrake** [reply]
Also if you love the branding of MS365 and Entra be aware that Microsoft is in the process of changing all their branding to copilot even i you don't have a copilot license lol.  Because of course they are.

**Reply by loryanstrant** [reply]
> Also if you love the branding of MS365 and Entra be aware that Microsoft is in the process of changing all their branding to copilot even i you don't have a copilot license lol.  Because of course they are.

Yeah, I don't acknowledge that as a thing because it's universally accepted as stupid and confusing - even by many of the MS folks I work with.

There's about 1,000 MVPs (including myself) rolling into Redmond late March - you can bet we'll be carrying pitchforks on that topic. :-)

**Reply by RogerSelwyn** [reply]
> I do have a P1 license so if you need me to test something let me know. I also have a tenant with a windows premium account which includes a P2 license.

I think when I wrote this I was pissed off at MS. I setup and pay for a P1 license (£5.68/month), to be honest I have no idea whether I need it or not. I thought I had setup MFA, but it never asks me for it when I log into Entra, probably because I'm using full admin (yep I know I shouldn't, but when I get charged for each MS365 account and I only have 3, I don't need the uplift).

It nags me when I log into Entra saying that MFA is needed, and in line with MS standard the documentation leads you through several pages with no clarity on what is needed. I imagine you guys can confirm whether I need it.

**Reply by RogerSelwyn** [reply]
> Also if you love the branding of MS365 and Entra be aware that Microsoft is in the process of changing all their branding to copilot even i you don't have a copilot license lol. Because of course they are.

I recently change the branding on the integrations and was aware of the Copilot change. These integration are unlikely to ever do anything Copilot, so whilst I'll try to follow branding I wouldn't plan to change to anything that implies any level of intelligence.

**Reply by RogerSelwyn** [reply]
> Yeah, I don't acknowledge that as a thing because it's universally accepted as stupid and confusing - even by many of the MS folks I work with.
> 
> There's about 1,000 MVPs (including myself) rolling into Redmond late March - you can bet we'll be carrying pitchforks on that topic. :-)

They seem to have dumped Copilot onto everything. I'm getting too old I think, I can be doing with VS Code trying to shove it down my throat and I just tuned my MS365 Office Apps down to Family Classic because I don't want to pay £25 extra for something I will be very unlikely to ever use.

**Reply by loryanstrant** [reply]
> > Also if you love the branding of MS365 and Entra be aware that Microsoft is in the process of changing all their branding to copilot even i you don't have a copilot license lol. Because of course they are.
> 
> I recently change the branding on the integrations and was aware of the CoPlilot change. These integration are unlikely to ever do anything CoPilot, so whilst I'll try to follow branding I wouldn't plan to change to anything that implies any level of intelligence.

https://www.letmecorrectthatforyou.com/#copilot

**Reply by RogerSelwyn** [reply]
> https://www.letmecorrectthatforyou.com/#copilot

I stand corrected. It's still nonsense.

**Comment by dreamworks** [comment]
I would love to chat/discuss with you about the O365 ./. M365 challenge.
(done more than 15 years of SharePoint O365/C#/REST development)

**Reply by RogerSelwyn** [reply]
Happy to have a chat some time. Give me a ping on firstname @ surname org uk

**Comment by pantherale0** [comment]
Hi @RogerSelwyn - some great work here and I use mainly the Teams side of things, but have an awful lot of experience with Graph API itself.

As we're using OAuth flows, shouldn't the integrations use Application Credentials in HA as documented here? https://developers.home-assistant.io/docs/core/platform/application_credentials/

As a side note - breaking this down will help me contribute to areas I use the most (Teams), so will probably start sending some PRs soon for new functionality.

**Reply by RogerSelwyn** [reply]
I’ve looked at using the HA app credentials mechanism, but really the library the integrations use (python-o365) isn’t set up to work with the ha mechanism. HA wants to do the oauth calls itself, whilst the library is designed to handle it all itself.

**Reply by RogerSelwyn** [reply]
Any prs would be appreciated, just give me an idea what you plan before you submit something please.

**Comment by fenty17** [comment]
@RogerSelwyn - can you advise on the best way to upgrade from O365 to MS365 Calendar. Can we install side by side or save any config, or is it set the new one up from scratch?

**Reply by RogerSelwyn** [reply]
You can install side by side. If you have done config in calendars.yaml (or todos/tasks.yaml) then you will be able to copy that content across to MS365 after you install the new integration. After you have disabled O365, you could then rename any entities to match the old ones, avoiding having to update automations/etc.

I'm not sure it would be easy for me to automate it, because one is in yaml and the other is via UI (which will avoid the endless mis-config issues I get), but I'll chew on it before we get into any forced migration in the middle of next year.

**Reply by RogerSelwyn** [reply]
Just to add, I'm working through create an o365 service that will create the config entries for the relevant MS365 integrations. All you will need to do is click on the integration option to re-configure and then re-authenticate to MS.

**Comment by leedrake** [comment]
Roger,

This is a duplicate of the post on the HASS Forums, so feel free to respond in either - I'll get notified.

thank you so much for your hard work on this module.

I got the original o365 working and that seemed to be relatively easy (I’m pretty familiar with graph and application permissions in Entra). I noticed you were moving to separate installs using the new configuration and one per function. I think that’s great, and downloaded and installed the MS365 - Calendar module and got THAT working as well, and removed the yaml entries for the o365 module and am using just the new one.

A couple of possible edits for your documentation. When searching in HACS for the repository, since it’s not in the main one and you need to use custom the steps are:

Search for rogerselwyn/MS365-Calendar to find the repository when using custom. Without your name at the beginning it can’t find it.
Download the repository
Once the repository is downloaded into HACS you need to find it and choose install. I am not super familiar with HACS so this step wasn’t clear to me, but I eventually figured it out. I suppose I should have read the HACS documentation more thoroughly - this isn’t on you lol.
I installed and got it up and running with alt-login, I was having issues getting the regular one to work on o365 module and the alt worked fine, so I just switched to that and continued using it for the ms365 module.

I have just a couple questions:

If I am loading the calendar for more than one user into HASS do I just add it as a second integration with a different email account name? Or is there a method to add an additional user using the existing entry? If I am adding it again do I do that from HACS or from the HASS integration page?
If I am doing a second user, since it’s delegated permissions, can I just use the same AppId and secret but a different email? Or do I need to add a different secret? Or a new separate application entry altogether?
Sorry if these are dumb questions - just trying to get clarity. I’m totally behind your new efforts and it’s awesome that I can actually write back calendar, todo, etc. updates from HASS. Graph is so cool. I have VOIP phones and freepbx all over the house in an intercom system, so being able to send TTS Voice announcements to them when events are upcoming will be way cool :slight_smile:

Oh and just one more note - if you have entraID P1 or 2 and have the ability to configure conditional access it’s a REALLY good idea to set a policy that only lets the public IP address of your home server access the app you configure in Entra, and limit access to only accounts that you want to retrieve calendars for. If you have a fixed IP you can use that address as the allowed location, and block all other locations. This severely limits the ability for anything else to access it - even if they somehow got your app id and secret.

If you’d like to add a documentation section on doing that, I’d be happy to write it. Let me know.

**Reply by RogerSelwyn** [reply]
> A couple of possible edits for your documentation. When searching in HACS for the repository, since it’s not in the main one and you need to use custom the steps are:
> 
> Search for rogerselwyn/MS365-Calendar to find the repository when using custom. Without your name at the beginning it can’t find it. Download the repository Once the repository is downloaded into HACS you need to find it and choose install. I am not super familiar with HACS so this step wasn’t clear to me, but I eventually figured it out. I suppose I should have read the HACS documentation more thoroughly - this isn’t on you lol. I installed and got it up and running with alt-login, I was having issues getting the regular one to work on o365 module and the alt worked fine, so I just switched to that and continued using it for the ms365 module.

To be honest it will be on HACS Default in 3/4 weeks, so it will be found in the standard way via HACS.

> I have just a couple questions:
> 
> If I am loading the calendar for more than one user into HASS do I just add it as a second integration with a different email account name? Or is there a method to add an additional user using the existing entry? If I am adding it again do I do that from HACS or from the HASS integration page? If I am doing a second user, since it’s delegated permissions, can I just use the same AppId and secret but a different email? Or do I need to add a different secret? Or a new separate application entry altogether? Sorry if these are dumb questions - just trying to get clarity. I’m totally behind your new efforts and it’s awesome that I can actually write back calendar, todo, etc. updates from HASS. Graph is so cool. I have VOIP phones and freepbx all over the house in an intercom system, so being able to send TTS Voice announcements to them when events are upcoming will be way cool :slight_smile:

Multiple questions in here:

- Just add as a second integration. The HA config UI is slightly unwieldy so adding two via the same integration just gave unnecessary complexity, and would only saves a couple of calls at startup. It would be the same number if calls going forward.
- HACS installs the code, HA creates an integration instance. So there would be nothing you could do on HACS to create a second instance.
- Yes you can use the same App ID and Secret, you can use the same one across the different integrations as well, if you so choose.

> Oh and just one more note - if you have entraID P1 or 2 and have the ability to configure conditional access it’s a REALLY good idea to set a policy that only lets the public IP address of your home server access the app you configure in Entra, and limit access to only accounts that you want to retrieve calendars for. If you have a fixed IP you can use that address as the allowed location, and block all other locations. This severely limits the ability for anything else to access it - even if they somehow got your app id and secret.
> 
> If you’d like to add a documentation section on doing that, I’d be happy to write it. Let me know.

I'm always open to a PR, and would consider including it. The challenge comes when people who aren't capable try to do it and then blame the integration for it not working. So I'd have to think about whether I'd get extra support requests or not. You just need to look at the number of issues raised where people simply haven't followed the instructions (such as using the ID of the client secret, rather than the value).

**Reply by loryanstrant** [reply]
> I'm always open to a PR, and would consider including it. The challenge comes when people who aren't capable try to do it and then blame the integration for it not working. So I'd have to think about whether I'd get extra support requests or not. You just need to look at the number of issues raised where people simply haven't followed the instructions (such as using the ID of the client secret, rather than the value).
> 
> 

Mate, you'd be surprised how many IT admins do this as well!

**Reply by RogerSelwyn** [reply]
> Mate, you'd be surprised how many IT admins do this as well!

But they'd probably figure it out for themselves rather than raising an issue. The challenge is that the errors that I get back through authentication don't necessarily allow me to be conclusive on what I provide to the user. Or else I just need to work harder...

**Reply by RogerSelwyn** [reply]
> > Mate, you'd be surprised how many IT admins do this as well!
> 
> But they'd probably figure it out for themselves rather than raising an issue. The challenge is that the errors that I get back through authentication don't necessarily allow me to be conclusive on what I provide to the user. Or else I just need to work harder...

Ahh, just looked and I have coded for that one :-)

**Reply by loryanstrant** [reply]
Nah, I mean I instruct an IT admin to create an app reg for me and to provide the tenant ID, client ID, and secret - and they give me the secret ID instead of value. Happens all too often, with people who *should* know what they're doing. :-/

**Reply by pantherale0** [reply]
I've found a lot of the time IT admins don't really know what the Graph API does exactly and what the scopes do.

Spent months battling a long time ago around delegated vs application scopes. Even with people who specialise in authentication

**Reply by leedrake** [reply]
I am so with both of you on this, and you're right, unless the person has a business premium account or Entra P1/P2 they won't be able to do conditional access. I do appreciate not adding complexity that the average user (and occasional sysadmin) might not get.

**Reply by leedrake** [reply]
Thank you for the succinct answers those completely answered my questions.  I appreciate the time you took. If I get any other ideas or start to delve into the code for other things to try I'll let you know :)  As a 42 year IT support veteran (now mostly retired) I certainly get the not wanting to increase the support load thing - especially for a free volunteer code base.

**Reply by RogerSelwyn** [reply]
> I am so with both of you on this, and you're right, unless the person has a business premium account or Entra P1/P2 they won't be able to do conditional access. I do appreciate not adding complexity that the average user (and occasional sysadmin) might not get.

Can you comment as to whether I need a P1 account to logon to Entra? It seemed to imply I did. I set up conditional access for family/etc. If I just move to MFA across the board, thus removing conditional access, can I bin the P1 for me as admin?

**Reply by pantherale0** [reply]
@RogerSelwyn you don't need a P1 or P2 license to login to Entra. Or even use to use MFA. Then again I have a business premium subscription.

**Reply by RogerSelwyn** [reply]
> Thank you for the succinct answers those completely answered my questions. I appreciate the time you took. If I get any other ideas or start to delve into the code for other things to try I'll let you know :) As a 42 year IT support veteran (now mostly retired) I certainly get the not wanting to increase the support load thing - especially for a free volunteer code base.

I only made it to 36 years in IT, but definitely retired, I couldn't cope with anymore at my last company. So this helps to keep me busy, but I can't do with adding complexity for those that can't handle it. See my question back to you on MFA/P1, if it confuses me then it sure as hell will confuse those who don't work with it as a matter of course.

**Reply by RogerSelwyn** [reply]
> @RogerSelwyn you don't need a P1 or P2 license to login to Entra. Or even use to use MFA. Then again I have a business premium subscription.

Not sure why it kept telling me I did when I logged in then. Came up with a popup at top right (which isn't there now I log back in). I probably ought to just remove the MFA conditional access and remove the subscription. Need to do it when I won't screw up family access.

**Reply by pantherale0** [reply]
You can check your licensed features here:
https://portal.azure.com/#view/Microsoft_AAD_IAM/LicensesMenuBlade/~/LicensedFeaturesState

For phone & sms based MFA, you need a O365/P1/P2 license for example. I've actually just looked over my own licenses and I have business standard, not premium

![image](https://github.com/user-attachments/assets/4f7219e3-0355-4e18-b077-1014bdc84a38)

**Reply by pantherale0** [reply]
Ahh I've just seen, conditional access is only available under P1/P2. Although it seems you can set it up with a trial license and the config will still continue to apply long after that trial has ended.

**Reply by RogerSelwyn** [reply]
Yep, but it costs, not a lot, but it provides me very little. If I change back to per-user, I can remove the conditional. At the moment it is set with no MFA on my admin account which is shocking practice (but how the walkthrough sets it up). I just need to fix it when I'm bored enough.

**Reply by leedrake** [reply]
Yeah Microsoft raised the price of P1 ridiculously.  If it were a buck a month or something fine, but at it's current price most won't bother to buy it.  It's really just a layer of security everyone should have access to. I hate their policy of "The cheaper it is the less secure it is".  If we just started with full security features and differentiated based on actual user features it would reduce so much hacking and compromise issues. Sorry to introduce this and the above discussion certainly illustrates why it might become a support headache. I withdraw my suggestion. 👍

**Reply by RogerSelwyn** [reply]
So I've just gone through and added MFA to my account, and will swap everyone to having MFA enabled/enforced. Some re-login will be required.

Good news, it's identified an exception in the code I need to handle better.

**Comment by leedrake** [comment]
yeah mfa is essential these days.  I try to avoid creating accounts on
services or sites that don't offer it.  Plus Microsoft has already made it
mandatory for admin, and will probably make it mandatory for all soon.
Might as well get the pain out of the way now.
-----------------------------------
Lee Drake
***@***.***

**Reply by RogerSelwyn** [reply]
That's the thing, it wasn't switched on for my admin, so it wasn't mandatory. It is enabled for my admin account now.

**Reply by pantherale0** [reply]
Very unlikely to see Microsoft force MFA for all users. That would be a complete headache for a lot of 1st line support you would then have to explain to every person that by installing the authenticator app or receiving a text message on your phone IT isn't spying on your personal life.

The only time MFA has been enforced on my admin accounts (both in work and outside of work) has been via conditional access policies.

**Reply by RogerSelwyn** [reply]
I think I agree on whether they will enforce or not. That said, with so few users, I'm just enabling per-user MFA on all accounts.

**Comment by Alexius30** [comment]
Hey,
I'm having a problem with authorization. I've followed all the steps in the documentation and everything has worked great so far. However, when linking to my MS account, there are problems with the URL that needs to be entered.
Error: It loads for several minutes until the message "Unknown error occurred" appears.
Does anyone have a solution for this?

**Reply by RogerSelwyn** [reply]
Please create a bug report with the requested information, including info from the system log.

**Comment by RogerSelwyn** [comment]
For those following this, you may have noticed that MS365 - Calendar, Mail and To Do have been added to HACS default recently, with Teams expected in a few weeks. Consequently I will be shutting down development on the O365 integration, apart from HA core mandated changes until 2025.11 when I believe the integration will fail (see the note on the Read Me).

I will add deprecation notices to the integration.

**Comment by RogerSelwyn** [comment]
Teams is now available on HACS default. Consequently there will be no new development work on O365.

**Comment by uSlackr** [comment]
I migrated last week with no issues.  Thanks for your continued work on this!


\\Greg

________________________________
